### DISCUSSION TITLE
Protect any route through middleware with custom sign-in page

### DISCUSSION THREAD

**Discussion by aspnetde** [nextauthjs/next-auth#5572] [Help]
Hi there. I've got a custom sign-in page configured in `[...nextauth].ts`:

```ts
export default NextAuth({
  // ...
  pages: {
    signIn: "/auth/signin",
  },
  // ...
});
```

According to the [docs](https://next-auth.js.org/configuration/nextjs#middleware) I now expected that configuring the same within the `middleware.ts` would work:

```ts
export { default } from "next-auth/middleware";

export const config = {
  pages: {
    signIn: "/auth/signin",
  },
};
```

My expectation: Now any route is protected and it recognizes the custom path to my sign-in page in case the user needs to be redirected there. But what actually happens is an endless redirect until the browser aborts, because it does not seem to recognize that custom route for the sign-in page.

My workaround now is to use the matcher, but I would sleep better if I knew that any route except for the auth stuff would be protected by default.

Did I make a mistake in configuring the middleware behavior, or could there be a bug?

Thanks
Thomas

**Comment by zacBkh** [comment]
Hello!

Not a next-auth expert but it seems what you are doing above is enabling page protection only for signIn page.
Are you trying to enable middleware everywehre except for signIn route ?

**Comment by johann-taberlet** [comment]
Hi,
I have the exact same question. I want to protect everything except my custom sign in /sign up pages.
I tried a lot of things with the matcher config but I'm not able to make it works (not an expert of regex).

My understanding is if we don't specify any matcher, every routes are protected. But since we add a negative matcher for our singin page, we have to handle every other routes in the matcher and I don't find the right regex.
What I tried so far to unprotect `/login` route:

```ts
export { default } from "next-auth/middleware";
export const config = {
  matcher: ["/((?!login))"], 
// unprotect "/login" but since there is an entry in the matcher, it unprotect every other routes
};
```
```ts
export const config = {
  matcher: ["/", "/((?!login))"], 
// unprotect "/login", protect "/" but unprotect "/something"
};
```
```ts
export const config = {
  matcher: ["/:path*", "/((?!login))"], 
// protect everything again, "/:path*" seems to cancel the second negative matcher
};
```

Any ideas on what is the right way of achieve this trivial thing ? (should be in the doc imo)

**Reply by zacBkh** [reply]
Hello!
Tbh I finally ended up using [the good old next js middleware feature](https://nextjs.org/docs/advanced-features/middleware)...

Overall, the idea is to list in an array all the path you want to restrict access to then handle (individually or not) each case in your middleware.
I am sure my below code can be optimized like crazy but it's working for now and I am in phase 1 of my project.

1.  **I created an array of protected paths** : 
`
const arrayOfProtectedPaths = ["/spots/newSpot", "/auth/profile"];
`

2.  **I referenced each paths in my middleware and handled individual cases** : 
_Please ignore some verbose related to query params and returnTo behaviour_
```
export async function middleware(req) {
    const pathname = req.nextUrl.pathname; // capture the page user requested access to
  
    const session = await getToken({
      req: req,
      secret: process.env.NEXTAUTH_SECRET,
    });
  
    // Protect protected pages
    if (arrayOfProtectedPaths.includes(pathname)) { // if the page the user is trying to access belongs to protected paths
      if (session === null) { // if user is not logged in
        const returnTo = req.nextUrl.pathname;
  
        if (pathname === "/auth/profile") {
          return NextResponse.redirect(
            `http://localhost:3008/auth/SignIn?mustLogIn=access your profile&returnTo=/auth/profile`
          );
        }
        if (pathname === "/spots/newSpot") {
          return NextResponse.redirect(
            `http://localhost:3008/auth/SignIn?mustLogIn=create a new spot&returnTo=${returnTo}`
          );
        }
      }
    }
}
```


Alternatively, there are [interesting discussions here.](https://github.com/nextauthjs/next-auth/discussions/5698)

Don't hesitate to keep in touch if you wish, I am a self taught web-dev (started one year ago)!
